### DISCUSSION TITLE
Using pdfbox in Mirth

### DISCUSSION THREAD

**Discussion by dougie-CIT** [nextgenhealthcare/connect#4622] [Q&A]
Thanks for looking !!

I am relatively new to mirth( using 3.11.0)  - have tried to paste the channel but renamed to txt as it doesnt accept xml)

[TEST_PDF_PARSER.txt](https://github.com/nextgenhealthcare/connect/files/6768700/TEST_PDF_PARSER.txt)

I have a channel I am trying to read PDF files from a folder and parse some text using pdfbox. I have the code parsing working fine elsewhere.

I have added  pdfbox -1.8.16.jar and fontbox 1.8.16.jar to the custom-lib folder and removed the pdfbox jar from the system-lib folder ( i dont need the pdf viewer in mirth)

Using this code in the transformer :

```
importPackage(org.apache.pdfbox.multipdf);
importPackage(org.apache.pdfbox.pdmodel);

var filepath = 'c:\\TESTHL7IN\\PDFIN\\' + $('pdfname');
logger.info(filepath);

var fbyte = FileUtil.readBytes( filepath);
logger.info(fbyte.length);
var pdfd = org.apache.pdfbox.pdmodel.PDDocument.load(new java.io.ByteArrayInputStream(fbyte));
```

I get this error :

```
Transformer error
ERROR MESSAGE: Error evaluating transformer
com.mirth.connect.server.MirthJavascriptTransformerException: 
CHANNEL:	EKORA_PDF_PARSER
CONNECTOR:	Destination 1
SCRIPT SOURCE:	TRANSFORMER
SOURCE CODE:	
106: var filepath = 'c:\\TESTHL7IN\\PDFIN\\' + $('pdfname');
107: logger.info(filepath);
108: 
109: var fbyte = FileUtil.readBytes( filepath);
110: logger.info(fbyte.length);
111: var pdfd = new org.apache.pdfbox.pdmodel.PDDocument();
112: pdfd.load(new java.io.ByteArrayInputStream(fbyte));
113: // var document = new org.apache.pdfbox.pdmodel.PDDocument.load(new java.io.ByteArrayInputStream(fbyte));
114: 
115: 
LINE NUMBER:	111
DETAILS:	TypeError: [JavaPackage org.apache.pdfbox.pdmodel.PDDocument] is not a function, it is object.
	at 90d72036-0b4f-4704-9653-6039a4d1204d:111 (doTransform)
	at 90d72036-0b4f-4704-9653-6039a4d1204d:149 (doScript)
	at 90d72036-0b4f-4704-9653-6039a4d1204d:151
	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer$FilterTransformerTask.doCall(JavaScriptFilterTransformer.java:154)
	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer$FilterTransformerTask.doCall(JavaScriptFilterTransformer.java:119)
	at com.mirth.connect.server.util.javascript.JavaScriptTask.call(JavaScriptTask.java:113)
	at java.util.concurrent.FutureTask.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)
```

**Comment by rjkroll** [comment]
What happens if you comment out lines 109,110,111 and 112 and use this:

var pdfd  = new org.apache.pdfbox.pdmodel.PDDocument.load(new java.io.ByteArrayInputStream(FileUtil.readBytes(filepath)));

**Reply by dougie-CIT** [reply]
Hi @rjkroll same error... I added the lines to check I was reading the file and it was correct size.

I assume im not calling the library properly in some way. I can see its loaded fine as a custom library.

**Comment by rjkroll** [comment]
Not sure then, it looks like there might be some issues with versioning and how Mirth works with custom-libs?

https://forums.mirthproject.io/forum/mirth-connect/support/12934-problems-with-pdfbox-in-mirthconnect-3-2-1

**Reply by dougie-CIT** [reply]
Thanks . yep i saw that and have tried all of these, but cant get it to work frustratingly

**Reply by pacmano1** [reply]
I think @tonygermano did a write up on this topic at https://forums.mirthproject.io/forum/mirth-connect/support/175423-javascript-error-using-apache-pdfbox-and-fontbox

**Reply by dougie-CIT** [reply]
@pacmano1  great thanks Ill have a look

**Reply by NoraTheExplorer** [reply]
> https://forums.mirthproject.io/forum/mirth-connect/support/12934-problems-with-pdfbox-in-mirthconnect-3-2-1

oh my, never thought I'd see a thread of mine as a reference somewhere :)
I'm still working with pdfbox-1.8.10 but it should work with .16 as well.

@dougie-CIT Anyhow, I read that you removed the pdfbox jar from the system-lib folder. That shouldn't be an issue.
I added the following files to the folder extensions/doc/lib

- fontbox-1.8.10.jar
- jempbox-1.8.10.jar
- pdfbox-1.8.10.jar
- preflight-1.8.10.jar
- xmpbox-1.8.10.jar

and edited the extensions/doc/destination.xml like this:
```
<connectorMetaData path="doc">
  <name>Document Writer</name>
  <author>NextGen Healthcare</author>
  <pluginVersion>3.10.1</pluginVersion>
  <mirthVersion>3.10.1</mirthVersion>
  <url>http://www.nextgen.com</url>
  <description>This connector allows Mirth Connect to create RTF or PDF documents based on an HTML template. PDF files can be password protected.</description>
  <clientClassName>com.mirth.connect.connectors.doc.DocumentWriter</clientClassName>
  <serverClassName>com.mirth.connect.connectors.doc.DocumentDispatcher</serverClassName>
  <sharedClassName>com.mirth.connect.connectors.doc.DocumentDispatcherProperties</sharedClassName>
  <library type="CLIENT" path="doc-client.jar" />
  <library type="SHARED" path="doc-shared.jar" />
  <library type="SERVER" path="doc-server.jar" />
  <library type="SERVER" path="lib/itext-2.1.7.jar" />
  <library type="SERVER" path="lib/itext-rtf-2.1.7.jar" />
  <library type="SERVER" path="lib/flying-saucer-core-9.0.1.jar" />
  <library type="SERVER" path="lib/flying-saucer-pdf-9.0.1.jar" />
  <apiProvider type="SERVLET_INTERFACE" name="com.mirth.connect.connectors.doc.DocumentConnectorServletInterface" />
  <apiProvider type="SERVER_CLASS" name="com.mirth.connect.connectors.doc.DocumentConnectorServlet" />
  <transformers>
  </transformers>
  <protocol>doc</protocol>
  <type>DESTINATION</type>
  <library type="SERVER" path="lib/fontbox-1.8.10.jar" />
  <library type="SERVER" path="lib/jempbox-1.8.10.jar" />
  <library type="SERVER" path="lib/pdfbox-1.8.10.jar" />
  <library type="SERVER" path="lib/preflight-1.8.10.jar" />
  <library type="SERVER" path="lib/xmpbox-1.8.10.jar" />
</connectorMetaData>
```

This just replaces the 1.8.4 that mirth uses with the 1.8.10. In your case the .16.

I use a different function to create the byteArrayInputStream.
filex contains a base64 encoded PDF. I also use, even if not recommended, the new command. Maybe this is the difference here.

```
bais = new java.io.ByteArrayInputStream(FileUtil.decode(filex))
pddoc = new Packages.org.apache.pdfbox.pdmodel.PDDocument.load(bais) // class from PDFBOX from apache
pages = pddoc.getDocumentCatalog().getAllPages() // method from PDFBOX from apache
```

**Reply by dougie-CIT** [reply]
@NoraTheExplorer  always nice to see your work recycling !!!! - its much appreciated !

Ive got it working as you suggest so thanks again !!

**Comment by dougie-CIT** [comment]
I now have the Resource set up as described in @tonygermano   write up - thanks @pacmano1 

and my code is as follows in the destination transformer :

`const iso = Packages(contextFactory.getIsolatedClassLoader());
var pdDocument = new iso.org.apache.pdfbox.pdmodel.PDDocument();

var filepath = 'c:\\TESTHL7IN\\PDFIN\\' + $('pdfname');
logger.info(filepath);

var fbyte = FileUtil.readBytes( filepath);
logger.info(fbyte.length);
var pdfd = pdDocument.load(new java.io.ByteArrayInputStream(fbyte));
`

This is my resource :

![image](https://user-images.githubusercontent.com/68617355/124639988-b2198b80-de84-11eb-8198-5156bfed5f05.png)

This is it linked in the channel :

![image](https://user-images.githubusercontent.com/68617355/124640252-12103200-de85-11eb-9c38-b466396360f2.png)

This is the error :

`Transformer error
ERROR MESSAGE: Error evaluating transformer
com.mirth.connect.server.MirthJavascriptTransformerException: 
CHANNEL:	TEST_PDF_PARSER
CONNECTOR:	Destination 1
SCRIPT SOURCE:	TRANSFORMER
SOURCE CODE:	
108: function doFilter() { phase[0] = 'filter'; return true; }function doTransform() { phase[0] = 'transformer'; logger = Packages.org.apache.log4j.Logger.getLogger(phase[0]);
109: 
110: 
111: // add this to the top of you script
112: const iso = Packages(contextFactory.getIsolatedClassLoader());
113: var pdDocument = new iso.org.apache.pdfbox.pdmodel.PDDocument();
114: 
115: var filepath = 'c:\\TESTHL7IN\\PDFIN\\' + $('pdfname');
116: logger.info(filepath);
117: 
LINE NUMBER:	113
DETAILS:	TypeError: [JavaPackage org.apache.pdfbox.pdmodel.PDDocument] is not a function, it is object.
	at 5f53f030-ed2a-4c3e-8efb-dc560dc02fdd:113 (doTransform)
	at 5f53f030-ed2a-4c3e-8efb-dc560dc02fdd:146 (doScript)
	at 5f53f030-ed2a-4c3e-8efb-dc560dc02fdd:148
	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer$FilterTransformerTask.doCall(JavaScriptFilterTransformer.java:154)
	at com.mirth.connect.server.transformers.JavaScriptFilterTransformer$FilterTransformerTask.doCall(JavaScriptFilterTransformer.java:119)
	at com.mirth.connect.server.util.javascript.JavaScriptTask.call(JavaScriptTask.java:113)
	at java.util.concurrent.FutureTask.run(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.lang.Thread.run(Unknown Source)
`

Thanks all for help - much appreciated

**Reply by tonygermano** [reply]
FYI, code blocks have triple back ticks and look like this.
````
```
some text
more test
```
````

The error indicates that it can't locate the PDDocument class when trying to construct a new instance, but the resource and channel dependency look to be set up correctly. Have you redeployed the channel since assigning the resource?

You don't actually need to create the new PDDocument in that line because `load` is a static method, but it still should have worked.

```js
const iso = Packages(contextFactory.getIsolatedClassLoader());

var filepath = 'c:\\TESTHL7IN\\PDFIN\\' + $('pdfname'); // note you need to escape your backslashes
logger.info(filepath);

// var fbyte = FileUtil.readBytes( filepath);
// logger.info(fbyte.length);

// static load method can take a string filename
var pdfd = iso.org.apache.pdfbox.pdmodel.PDDocument.load(filepath);
```

I also think that you do not need all of those jar files. I'm pretty sure that the "standalone" `app` version contains everything that's in both of the other jars.

**Reply by dougie-CIT** [reply]
Many thanks. Noted re the ticks!

I shall try to clear the other jars and have another go with just the app one as see if both is maybe causing the conflict when I get back tomorrow.

Thanks again for the advice @tonygermano

**Reply by dougie-CIT** [reply]
For info I could not get this to work, ended up just adding fontbox and going back to 1.8.4 for both which has met my needs. Ill try again once ive solved the immediate issue. Thanks again to all for help.

I used the method here : https://forums.mirthproject.io/forum/mirth-connect/support/12934-problems-with-pdfbox-in-mirthconnect-3-2-1

1. Copy your new libs into ...\extensions\doc\lib
2. open ...\extensions\doc\destination.xml with Notepad or another Texteditor
3. replace and/or add your libs in that file

**Reply by tonygermano** [reply]
sure, just know that your changes will need to be redone if you upgrade mirth.
